File: chord_builder.py
from sys import stdin
from pprint import pprint
from os import system

class Tab(object):

    STRINGS = {
        6: [2,   2.5, 3,   3.5, 4,   4.5],
        5: [4.5, 5,   5.5, 0,   0.5, 1],
        4: [1,   1.5, 2,   2.5, 3,   3.5],
        3: [3.5, 4,   4.5, 5,   5.5, 0],
        2: [5.5, 0,   0.5, 1,   1.5, 2],
        1: [2,   2.5, 3,   3.5, 4,   4.5]
    }

    FRETS = {
        1: ['E', 'F', 'F#', 'G', 'G#', 'A'],
        2: ['B', 'C', 'C#', 'D', 'D#', 'E'],
        3: ['G', 'G#', 'A', 'A#', 'B', 'C'],
        4: ['D', 'D#', 'E', 'F', 'F#', 'G'],
        5: ['A', 'A#', 'B', 'C', 'C#', 'D'],
        6: ['E', 'F', 'F#', 'G', 'G#', 'A']
    }

    NOTES = {
        0:   'C',
        0.5: 'C#', 
        1:   'D',
        1.5: 'D#',
        2:   'E',
        2.5: 'F',
        3:   'F#',
        3.5: 'G',
        4:   'G#',
        4.5: 'A',
        5:   'A#',
        5.5: 'B'
    }

    frets = {
        1: -1,
        2: -1,
        3: -1,
        4: -1,
        5: -1,
        6: -1
    }

    def read_frets(self):
        i = 1
        while i < 7:
            print('enter fret # for line ' + str(i) + ': ', flush=True, end='')
            fret_str = stdin.readline().strip()
            if fret_str == 'x':
                self.frets[i] = -1
                i += 1
            elif fret_str.isnumeric() and (int(fret_str) >= 0 and int(fret_str) <= 5):
                self.frets[i] = int(fret_str)
                i += 1
        #pprint(self.frets)

    def draw_frets(self):
        system('cls')
        print('')
        for i in range(1, 7):
            if self.frets[i] == -1:
                out_str = 'x'
            else:
                out_str = str(self.frets[i])
            out_str += '╟'
            for f in range(1, 6):
                if self.frets[i] == f:
                    out_str += '─O─┼'
                else:
                    out_str += '───┼'
            if self.frets[i] == -1:
                pass
            else:
                out_str += self.FRETS[i][self.frets[i]]
            print(out_str)

    def build_intervals(self):
        notes = []
        chords = []
        for i in range(1, 7):
            if self.frets[i] == -1:
                continue
            else:
                notes.append(self.STRINGS[i][self.frets[i]])
        #print('build_intervals(): notes ', notes)
        for j in notes:
            ints_tmp = []
            types = {}
            for i in notes:
                diff = i - j
                if diff < 0:
                    diff += 6
                #print('build_intervals(): current note:', self.NOTES[j], 'note:', self.NOTES[i], 'diff:', diff)
                if diff == 0:
                    continue
                ints_tmp.append(diff)
            intervals = set(sorted(ints_tmp))
            #print('build_intervals(): root', self.NOTES[j], intervals)
            chord_name = self.NOTES[j]
            types['V']   = self.type_V(intervals)
            types['III'] = self.type_III(intervals)
            chord_name += types['III']
            chord_name += types['V']
            chord_name += self.is_7(intervals, types)
            chord_name += self.is_9(intervals, types)
            chord_name += self.is_11(intervals, types)
            chord_name += self.is_6(intervals, types)
            chords.append(chord_name)
        return sorted(set(chords))

    def type_V(self, intervals):
        ''' возвращает тип по ступени V - доминантный, уменьшенный, увеличенный '''
        if 3.5 in intervals and (3.0 not in intervals or 4.0 not in intervals):
            intervals.remove(3.5)
            return ''
        if 3.0 in intervals and (3.5 not in intervals or 4.0 not in intervals):
            intervals.remove(3.0)
            return '/dim5'
        if 4.0 in intervals and (3.0 not in intervals or 3.5 not in intervals):
            intervals.remove(4.0)
            return '/aug5'
        return '(No5)'

    def type_III(self, intervals):
        ''' возвращает тип по ступени III - мажор, минор, sus2, sus4 '''
        if 1.5 in intervals: # and (1.0 not in intervals and 2.0 not in intervals and 2.5 not in intervals):
            intervals.remove(1.5)
            return 'm'
        if 2.0 in intervals: # and (1.0 not in intervals and 1.5 not in intervals and 2.5 not in intervals):
            intervals.remove(2.0)
            return ''
        if 1.0 in intervals: # and (1.5 not in intervals and 2.0 not in intervals and 2.5 not in intervals):
            intervals.remove(1.0)
            return 'sus2'
        if 2.5 in intervals: # and (1.0 not in intervals and 1.5 not in intervals and 2.0 not in intervals):
            intervals.remove(2.5)
            return 'sus4'
        return '(No3)'
            
    def is_9(self, intervals, types): 
        ''' только если есть III '''
        if types['III'] == '(No3)':
            return ''
        if 0.5 in intervals and 1.0 not in intervals:
            intervals.remove(0.5)
            return '/b9'
        if 1.0 in intervals and 0.5 not in intervals:
            intervals.remove(1.0)
            return '/9'
        return ''

    def is_11(self, intervals, types):
        ''' только если есть III '''
        if types['III'] == '(No3)':
            return ''
        if 2.5 in intervals:
            intervals.remove(2.5)
            return '/11'
        return ''

    def is_6(self, intervals, types):
        if 4.0 in intervals and 4.5 not in intervals and types['V'] != 'Aug5':
            intervals.remove(4.0)
            return '/b6'
        if 4.5 in intervals and 4.0 not in intervals:
            intervals.remove(4.5)
            return '/6'
        return ''

    def is_7(self, intervals, types):
        if 4.5 in intervals and (5.0 not in intervals and 5.5 not in intervals):
            intervals.remove(4.5)
            return '/b7'
        if 5.0 in intervals and (4.5 not in intervals and 5.5 not in intervals):
            intervals.remove(5.0)
            return '/7'
        if 5.5 in intervals and (4.5 not in intervals and 5.0 not in intervals):
            intervals.remove(5.5)
            return '/maj7'
        return ''



if __name__ == '__main__':
    tab = Tab()
    tab.read_frets()
    tab.draw_frets()
    print('\n', tab.build_intervals())

File: chord_builder.js
"use strict";

// it runs on windows scripting host.

var STRINGS = new Array(6);
for (var i = 0; i < 6; i++) {
    STRINGS[i] = new Array(6);
}

STRINGS[0] = new Array(2.0, 2.5, 3.0, 3.5, 4.0, 4.5);
STRINGS[1] = new Array(5.5, 0.0, 0.5, 1.0, 1.5, 2.0);
STRINGS[2] = new Array(3.5, 4.0, 4.5, 5.0, 5.5, 0.0);
STRINGS[3] = new Array(1.0, 1.5, 2.0, 2.5, 3.0, 3.5);
STRINGS[4] = new Array(4.5, 5.0, 5.5, 0.0, 0.5, 1.0);
STRINGS[5] = new Array(2.0, 2.5, 3.0, 3.5, 4.0, 4.5);

var FRETS = new Array(6);
for (var i = 0; i < 6; i++) {
    FRETS[i] = new Array(6);
}

FRETS[0] = new Array("E", "F", "F#", "G", "G#", "A");
FRETS[1] = new Array("B", "C", "C#", "D", "D#", "E");
FRETS[2] = new Array("G", "G#", "A", "A#", "B", "C");
FRETS[3] = new Array("D", "D#", "E", "F", "F#", "G");
FRETS[4] = new Array("A", "A#", "B", "C", "C#", "D");
FRETS[5] = new Array("E", "F", "F#", "G", "G#", "A");

var NOTES = new Object({
0.0: "C",
0.5: "C#",
1.0: "D",
1.5: "D#",
2.0: "E",
2.5: "F",
3.0: "F#",
3.5: "G",
4.0: "G#",
4.5: "A",
5.0: "A#",
5.5: "B"});

//************************ main() ***********************//

var frets = readFrets();
drawFrets(frets);
var notes = fillNotes(frets);
var intervals = buildIntervals(notes);
var chords    = buildChords(intervals, frets).sort();
for (var i = 0; i < chords.length; i++) {
    if (frets[i] == -1) {
        continue;
    }
    if (chords[i+1] == chords[i]) {
        continue;
    } else {
        alert(chords[i]);
    }
}

//********************* functions() *********************//

function readFrets() {
    var i = 0;
    var frets = new Array();
    while (i < 6) {
        WScript.StdOut.write("Enter fret # for string #" + i + ":");
        var input = WScript.StdIn.readline();
        var inputInt = parseInt(input);
        if (input.toLowerCase() == "x") {
            frets.push(-1);
        } else if ((inputInt >= 0) && (inputInt <= 5)) {
            frets.push(inputInt);
        } else {
            continue;
        }
        i++;
    }
    return frets;
}

function fillNotes(frets) {
    var notes = new Array(); 
    for (var i = 0; i < 6; i++) {
        if (frets[i] == "x") {
            notes.push(-1);
        } else {
            notes.push(STRINGS[i][frets[i]]);
        }
    }
    return notes;
}

function buildIntervals(notes) {
    var intervals = new Array();
    for (var i = 0; i < 6; i++) {
        intervals[i] = new Array();
        for (var j = 0; j < 6; j++) {
            interval = notes[j] - notes[i];
            if (notes[j] == -1) {
                continue;
            }
            if (interval < 0) {
                interval += 6;
            }
            if (isInArray(intervals[i], interval) == false) {
                intervals[i].push(interval);
            }
        }
    }
    return intervals;
}

function drawFrets(frets) {
    var out_str = "\n";
    for(var i = 0; i < 6; i++) {
        if (frets[i] == -1) {
            out_str += "x";
        } else {
            out_str += frets[i];
        }
        out_str += "|";
        for (var j = 1; j < 6; j++) {
            if (frets[i] == j) {
                out_str += "-O-|";
            } else {
                out_str += "---|";
            }
        }
        out_str += "\n";
    }
    alert(out_str);
}

function isInArray(array, interval_value) {
    for (var k = 0; k < array.length; k++) {
        if (array[k] == interval_value) {
            return true;
        }
    }
    return false;
}

function buildChords(intervals, frets) {
    var iSet, chordName;
    var chords = new Array();
    var chordTypes = new Object();
    chordTypes = ({"II":  "",
                   "III": "",
                   "IV":  "",
                   "V":   "",
                   "VI":  "",
                   "VII": ""});
    for (var i = 0; i < 6; i++) {
        iSet = intervals[i].sort();
        chordName = FRETS[i][frets[i]];
        //for (var j = 0; j < iSet.length; j++) { alert(iSet); }    
        chordTypes["III"] = is_3(iSet);
        chordTypes["V"]   = is_5(iSet);
        chordTypes["VII"] = is_7(iSet);
        chordTypes["VI"]  = is_6(iSet, chordTypes);
        chordTypes["II"]  = is_9(iSet, chordTypes);
        chordTypes["IV"]  = is_11(iSet, chordTypes);

        chordName += chordTypes["III"] 
                   + chordTypes["V"] 
                   + chordTypes["VII"]
                   + chordTypes["VI"]
                   + chordTypes["II"]
                   + chordTypes["IV"];
        chords.push(chordName);
    }
    return chords;
}

function is_5(iSet) {
    if ((isInArray(iSet, 3.5) == true) && (isInArray(iSet, 3.0) == false) && (isInArray(iSet, 4.0) == false)) {
        return "";
    }
    if ((isInArray(iSet, 3.0) == true) && (isInArray(iSet, 3.5) == false) && (isInArray(iSet, 4.0) == false)) {
        return "dim5";
    }
    if ((isInArray(iSet, 4.0) == true) && (isInArray(iSet, 3.0) == false) && (isInArray(iSet, 3.5) == false)) {
        return "aug5";
    }
    return "(No5)";
}

function is_3(iSet) {
    if ((isInArray(iSet, 2.0) == true) && (isInArray(iSet, 1.5) == false) && (isInArray(iSet, 2.5) == false)) { //&& (isInArray(iSet, 1.0) == false) 
        return ""; // major
    }
    if ((isInArray(iSet, 1.5) == true) && (isInArray(iSet, 2.0) == false) && (isInArray(iSet, 2.5) == false)) { //&& (isInArray(iSet, 1.0) == false) 
        return "m"; // minor
    }
    if ((isInArray(iSet, 1.0) == true) && (isInArray(iSet, 1.5) == false) && (isInArray(iSet, 2.0) == false) && (isInArray(iSet, 2.5) == false)) {
        return "sus2";
    }
    if ((isInArray(iSet, 2.5) == true) && (isInArray(iSet, 1.0) == false) && (isInArray(iSet, 2.0) == false) && (isInArray(iSet, 1.5) == false)) {
        return "sus4";
    }
    return "(No3)";
}

function is_7(iSet) {
    if (isInArray(iSet, 4.5) == true) {
        return "/b7";
    }
    if (isInArray(iSet, 5.0) == true) {
        return "/7";
    }
    if (isInArray(iSet, 5.5) == true) {
        return "/maj7";
    }
    return "";
}

function is_9(iSet, chordTypes) {
    if (isInArray(iSet, 0.5) == true) {
        return "/b9";
    }
    if (isInArray(iSet, 1.0) == true) {
        if (chordTypes["III"].search(/sus2/) < 0) {
            return "/9";
        }
    }
    return "";
}

function is_6(iSet, chordTypes) {
    if (isInArray(iSet, 4.0) == true) {
        if (chordTypes["V"].search(/aug5/) < 0) {
            return "/6";
        }
    }
    return "";
}

function is_11(iSet, chordTypes) {
    if (isInArray(iSet, 2.5) == true) {
        if (chordTypes["III"].search(/sus4/) < 0) {
            return "/11";
        }
    }
    return "";
}

// aka console.log():
function alert(message) {
    WScript.StdOut.writeline(message);
}
